firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Simulator.cpp
+ *
+ *  Created on: May 30, 2021
+ *      Author: Rohan
+ *
+ *  Simulator for power management
+ */
+
+#include "Pufferfish/Driver/Power/Simulator.h"
+
+#include <cmath>
+#include <random>
+
+#include "Pufferfish/Util/Timeouts.h"
+
+namespace Pufferfish::Driver::Power {
+
+// NOLINTNEXTLINE(cert-msc51-cpp)
+std::ranlux24_base prng;
+// NOLINTNEXTLINE(readability-magic-numbers)
+std::uniform_int_distribution<int> uniform_int(1, 100);
+static constexpr float uniform_width = 1;
+std::uniform_real_distribution<float> uniform_centered(-uniform_width / 2, uniform_width / 2);
+
+void Simulator::input_clock(uint32_t current_time) {
+  if (initial_time_ == 0) {
+    initial_time_ = current_time;
+  }
+  if (update_needed()) {
+    previous_time_ = current_time_;
+  }
+  current_time_ = current_time - initial_time_;
+}
+
+void Simulator::transform(uint32_t current_time, PowerManagement &power_management) {
+  input_clock(current_time);
+  if (charging_) {
+    transform_charge(power_management);
+  } else {
+    initial_time_ = current_time;
+    transform_discharge(power_management);
+  }
+}
+
+void Simulator::transform_charge(PowerManagement &power_management) {
+  if (!update_needed()) {
+    return;
+  }
+  power_management.charging = true;
+  power_management.power_left +=
+      (1 + (power_responsiveness * uniform_centered(prng))) / uniform_int(prng);
+  if (power_management.power_left >= max_charge) {
+    power_management.power_left = max_charge;
+    charging_ = false;
+  }
+}
+
+void Simulator::transform_discharge(PowerManagement &power_management) {
+  if (!update_needed()) {
+    return;
+  }
+  power_management.charging = false;
+  power_management.power_left -=
+      (1 + (power_responsiveness * uniform_centered(prng))) / uniform_int(prng);
+  if (power_management.power_left == 0) {
```

Comment:
If `transform_discharge` should be symmetric with `transform_charge`, should this `if` check be `power_management.power_left <= 0`? And should the body of the if branch also set `power_management.power_left = 0?

Reply:
I had done the same previously, although `cppcheck` complaints about `Core/Src/Pufferfish/Driver/Power/Simulator.cpp:53:35: style: Checking if unsigned expression 'power_management.power_left' is less than zero. [unsignedLessThanZero]
`

Reply:
Ah, ok - I had thought `power_management.power_left` was a `float` rather than a `uint*_t`. Do you think we should change it to `float`, or leave it as `uint*_t`?

Reply:
Let's change it to `float`?